File: app/src/main/java/com/nikola/jakshic/truesight/viewModel/HeroViewModel.java
package com.nikola.jakshic.truesight.viewModel;

import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import com.nikola.jakshic.truesight.model.Hero;
import com.nikola.jakshic.truesight.repository.HeroRepository;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import javax.inject.Inject;

public class HeroViewModel extends ViewModel {

    private static final String LOG_TAG = HeroViewModel.class.getSimpleName();

    private MutableLiveData<List<Hero>> list;
    private HeroRepository repository;
    private MutableLiveData<Boolean> loading;
    private boolean initialFetch;

    @Inject
    public HeroViewModel(HeroRepository repository) {
        this.repository = repository;
        list = new MutableLiveData<>();
        loading = new MutableLiveData<>();
        loading.setValue(false);
    }

    public void initialFetch(long id) {
        if (!initialFetch) {
            repository.fetchHeroes(list, loading, id);
            initialFetch = true;
        }
    }

    public void fetchHeroes(long id) {
        repository.fetchHeroes(list, loading, id);
    }

    public void sort(Comparator<Hero> comparator) {
        if (list.getValue() == null) return;
        List<Hero> sortedList = new ArrayList<>(list.getValue());
        Collections.sort(sortedList, comparator);
        list.setValue(sortedList);
    }

    public MutableLiveData<List<Hero>> getHeroes() {
        return list;
    }

    public MutableLiveData<Boolean> isLoading() {
        return loading;
    }
}

File: app/src/main/java/com/nikola/jakshic/dagger/ui/profile/heroes/HeroViewModel.kt
package com.nikola.jakshic.dagger.ui.profile.heroes

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.nikola.jakshic.dagger.repository.HeroRepository
import com.nikola.jakshic.dagger.ui.ScopedViewModel
import com.nikola.jakshic.dagger.ui.Status
import com.nikola.jakshic.dagger.vo.Hero
import kotlinx.coroutines.launch
import javax.inject.Inject

class HeroViewModel @Inject constructor(
        private val repository: HeroRepository) : ScopedViewModel() {

    lateinit var list: LiveData<List<Hero>>
        private set

    private val _status = MutableLiveData<Status>()
    val status: LiveData<Status>
        get() = _status

    private var initialFetch = false

    private val onSuccess: () -> Unit = { _status.value = Status.SUCCESS }
    private val onError: () -> Unit = { _status.value = Status.ERROR }

    fun initialFetch(id: Long) {
        if (!initialFetch) {
            initialFetch = true
            fetchHeroes(id)
            sortByGames(id)
        }
    }

    fun fetchHeroes(id: Long) {
        launch {
            _status.value = Status.LOADING
            repository.fetchHeroes(id, onSuccess, onError)
        }
    }

    fun sortByGames(id: Long) {
        list = repository.getHeroesLiveDataByGames(id)
    }

    fun sortByWinRate(id: Long) {
        list = repository.getHeroesLiveDataByWinrate(id)
    }

    fun sortByWins(id: Long) {
        list = repository.getHeroesLiveDataByWins(id)
    }

    fun sortByLosses(id: Long) {
        list = repository.getHeroesLiveDataByLosses(id)
    }
}